File: apps/nomad-browser-macos/NomadBrowser/NomadBrowser/DisplayList.swift
//
//  DisplayList.swift
//  NomadBrowser
//
//  Swift types for display list deserialization
//

import Foundation

/// A display list containing drawable items
struct DisplayList: Codable {
    let width: Float
    let height: Float
    let items: [DisplayItem]
}

/// A single drawable item
struct DisplayItem: Codable {
    let kind: DisplayItemKind
    let bounds: Rect
}

/// Type of display item
enum DisplayItemKind: Codable {
    case text(TextItem)
    
    private enum CodingKeys: String, CodingKey {
        case type
        case content
        case fontSize = "font_size"
        case isLink = "is_link"
        case linkUrl = "link_url"
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        
        // The Rust enum is serialized as a struct with "Text" variant
        let content = try container.decode(String.self, forKey: .content)
        let fontSize = try container.decode(Float.self, forKey: .fontSize)
        let isLink = try container.decode(Bool.self, forKey: .isLink)
        let linkUrl = try container.decodeIfPresent(String.self, forKey: .linkUrl)
        
        self = .text(TextItem(
            content: content,
            fontSize: fontSize,
            isLink: isLink,
            linkUrl: linkUrl
        ))
    }
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        
        switch self {
        case .text(let item):
            try container.encode(item.content, forKey: .content)
            try container.encode(item.fontSize, forKey: .fontSize)
            try container.encode(item.isLink, forKey: .isLink)
            try container.encodeIfPresent(item.linkUrl, forKey: .linkUrl)
        }
    }
}

/// Text display item
struct TextItem: Codable {
    let content: String
    let fontSize: Float
    let isLink: Bool
    let linkUrl: String?
}

/// A rectangle
struct Rect: Codable {
    let x: Float
    let y: Float
    let width: Float
    let height: Float
    
    func contains(x: Float, y: Float) -> Bool {
        return x >= self.x && x <= self.x + self.width &&
               y >= self.y && y <= self.y + self.height
    }
}

File: crates/nomad-core/src/display_list.rs
//! Display list for rendering.
//!
//! Contains structures for representing visual output that can be
//! serialized across the FFI boundary.

use serde::{Deserialize, Serialize};

/// A display list containing all drawable items.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DisplayList {
    /// Width of the viewport
    pub width: f32,
    /// Total height of content
    pub height: f32,
    /// List of items to draw
    pub items: Vec<DisplayItem>,
}

impl DisplayList {
    /// Creates a new empty display list.
    pub fn new(width: f32) -> Self {
        Self {
            width,
            height: 0.0,
            items: Vec::new(),
        }
    }

    /// Adds an item to the display list.
    pub fn add_item(&mut self, item: DisplayItem) {
        // Update height based on item bounds
        let item_bottom = item.bounds.y + item.bounds.height;
        if item_bottom > self.height {
            self.height = item_bottom;
        }
        self.items.push(item);
    }

    /// Serializes the display list to bytes (JSON format for cross-language compatibility).
    pub fn to_bytes(&self) -> Result<Vec<u8>, String> {
        serde_json::to_vec(self).map_err(|e| format!("Failed to serialize: {}", e))
    }

    /// Deserializes a display list from bytes (JSON format).
    pub fn from_bytes(bytes: &[u8]) -> Result<Self, String> {
        serde_json::from_slice(bytes).map_err(|e| format!("Failed to deserialize: {}", e))
    }
}

/// A single drawable item in the display list.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DisplayItem {
    /// Type of item
    pub kind: DisplayItemKind,
    /// Bounding box
    pub bounds: Rect,
}

/// Type of display item.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum DisplayItemKind {
    /// Text to draw
    Text {
        /// The text content
        content: String,
        /// Font size
        font_size: f32,
        /// Whether this is a link
        is_link: bool,
        /// URL if this is a link
        link_url: Option<String>,
    },
}

/// A rectangle.
#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
pub struct Rect {
    pub x: f32,
    pub y: f32,
    pub width: f32,
    pub height: f32,
}

impl Rect {
    /// Creates a new rectangle.
    pub fn new(x: f32, y: f32, width: f32, height: f32) -> Self {
        Self {
            x,
            y,
            width,
            height,
        }
    }

    /// Checks if a point is inside this rectangle.
    pub fn contains(&self, px: f32, py: f32) -> bool {
        px >= self.x && px <= self.x + self.width && py >= self.y && py <= self.y + self.height
    }
}
